wgpu-core/src/device/resource.rs

```diff
@@ -3406,3 +3413,55 @@ impl<A: HalApi> Resource<DeviceId> for Device<A> {
         &mut self.info
     }
 }
+
+pub struct SnatchableMarker;
+/// A guard that provides read access to snatchable data.
+pub type SnatchGuard<'a> = RwLockReadGuard<'a, SnatchableMarker>;
+/// A guard that allows snatching the snatchable data.
+pub type ExclusiveSnatchGuard<'a> = RwLockWriteGuard<'a, SnatchableMarker>;
+
+/// A value that is mostly immutable but can be "snatched" if we need to destroy
+/// it early.
+///
+/// In order to safely access the underlying data, the device's global snatchable
+/// lock must be taken. To guarentee it, methods take a read or write guard of that
+/// special lock.
+pub struct Snatchable<T> {
```

Comment:
Could we move the implementation of Snatchable (and the various associated types) into it's own module.